File: research/experiments/sweep/runs/opt-small-python-sonnet45/src/nelder_mead.py
"""nelder-mead: Derivative-free simplex optimizer.

Maintains n+1 vertices in n dimensions. At each step, replaces the worst vertex
via reflection, expansion, contraction, or shrinkage.
"""

import math
from typing import Callable, List, Optional
from result_types import (
    OptimizeOptions, OptimizeResult, default_options, convergence_message
)
from vec_ops import add, sub, scale, add_scaled, norm


def nelder_mead(
    f: Callable[[List[float]], float],
    x0: List[float],
    options: Optional[OptimizeOptions] = None
) -> OptimizeResult:
    """Minimize a function using the Nelder-Mead simplex algorithm.

    Args:
        f: Objective function to minimize
        x0: Initial point
        options: Optimization options (uses defaults if None)

    Returns:
        OptimizeResult with gradient=None and gradient_calls=0
    """
    if options is None:
        options = default_options()

    # Algorithm parameters
    alpha = 1.0   # Reflection coefficient
    gamma = 2.0   # Expansion coefficient
    rho = 0.5     # Contraction coefficient
    sigma = 0.5   # Shrink coefficient
    initial_simplex_scale = 0.05

    n = len(x0)
    function_calls = 0

    # Create initial simplex: vertex 0 = x0, vertex i = x0 + h*e_i
    simplex: List[List[float]] = [x0.copy()]
    for i in range(n):
        h = initial_simplex_scale * max(abs(x0[i]), 1.0)
        vertex = x0.copy()
        vertex[i] += h
        simplex.append(vertex)

    # Evaluate function at all vertices
    f_values = [f(v) for v in simplex]
    function_calls += n + 1

    iteration = 0

    while iteration < options.max_iterations:
        # Sort vertices by function value (ascending)
        indices = sorted(range(n + 1), key=lambda i: f_values[i])
        simplex = [simplex[i] for i in indices]
        f_values = [f_values[i] for i in indices]

        best = simplex[0]
        f_best = f_values[0]
        worst = simplex[n]
        f_worst = f_values[n]
        second_worst = simplex[n - 1]
        f_second_worst = f_values[n - 1]

        # Check convergence: function value spread (std dev)
        mean_f = sum(f_values) / (n + 1)
        std_f = math.sqrt(sum((fv - mean_f) ** 2 for fv in f_values) / (n + 1))

        if std_f < options.func_tol:
            return OptimizeResult(
                x=best,
                fun=f_best,
                gradient=None,
                iterations=iteration,
                function_calls=function_calls,
                gradient_calls=0,
                converged=True,
                message=convergence_message(type('ConvergenceReason', (), {'kind': 'function'})())
            )

        # Check convergence: simplex diameter
        diameter = 0.0
        for i in range(n + 1):
            for j in range(i + 1, n + 1):
                dist = norm(sub(simplex[i], simplex[j]))
                diameter = max(diameter, dist)

        if diameter < options.step_tol:
            return OptimizeResult(
                x=best,
                fun=f_best,
                gradient=None,
                iterations=iteration,
                function_calls=function_calls,
                gradient_calls=0,
                converged=True,
                message=convergence_message(type('ConvergenceReason', (), {'kind': 'step'})())
            )

        # Compute centroid of all vertices except worst
        centroid = [0.0] * n
        for i in range(n):
            centroid = add(centroid, simplex[i])
        centroid = scale(centroid, 1.0 / n)

        # Reflect worst through centroid
        reflected = add_scaled(centroid, sub(worst, centroid), -alpha)
        f_reflected = f(reflected)
        function_calls += 1

        # Accept reflection if between best and second-worst
        if f_best <= f_reflected < f_second_worst:
            simplex[n] = reflected
            f_values[n] = f_reflected
            iteration += 1
            continue

        # If reflection is best, try expansion
        if f_reflected < f_best:
            expanded = add_scaled(centroid, sub(reflected, centroid), gamma)
            f_expanded = f(expanded)
            function_calls += 1

            if f_expanded < f_reflected:
                simplex[n] = expanded
                f_values[n] = f_expanded
            else:
                simplex[n] = reflected
                f_values[n] = f_reflected
            iteration += 1
            continue

        # If reflection is worst, try contraction
        if f_reflected < f_worst:
            # Outside contraction
            contracted = add_scaled(centroid, sub(reflected, centroid), rho)
            f_contracted = f(contracted)
            function_calls += 1

            if f_contracted < f_reflected:
                simplex[n] = contracted
                f_values[n] = f_contracted
                iteration += 1
                continue
        else:
            # Inside contraction
            contracted = add_scaled(centroid, sub(worst, centroid), rho)
            f_contracted = f(contracted)
            function_calls += 1

            if f_contracted < f_worst:
                simplex[n] = contracted
                f_values[n] = f_contracted
                iteration += 1
                continue

        # Contraction failed, shrink all vertices toward best
        for i in range(1, n + 1):
            simplex[i] = add_scaled(best, sub(simplex[i], best), sigma)
            f_values[i] = f(simplex[i])
            function_calls += 1

        iteration += 1

    # Max iterations reached
    indices = sorted(range(n + 1), key=lambda i: f_values[i])
    best = simplex[indices[0]]
    f_best = f_values[indices[0]]

    return OptimizeResult(
        x=best,
        fun=f_best,
        gradient=None,
        iterations=iteration,
        function_calls=function_calls,
        gradient_calls=0,
        converged=False,
        message=convergence_message(type('ConvergenceReason', (), {'kind': 'maxIterations'})())
    )

File: research/experiments/sweep/runs/opt-small-kotlin-sonnet45/src/main/kotlin/optimization/NelderMead.kt
package optimization

import kotlin.math.max
import kotlin.math.abs
import kotlin.math.sqrt

/**
 * Nelder-Mead simplex optimizer (derivative-free).
 */

private const val ALPHA = 1.0  // Reflection coefficient
private const val GAMMA = 2.0  // Expansion coefficient
private const val RHO = 0.5    // Contraction coefficient
private const val SIGMA = 0.5  // Shrink coefficient
private const val INITIAL_SIMPLEX_SCALE = 0.05

private data class SimplexVertex(
    val x: DoubleArray,
    val fx: Double
) {
    fun copy(): SimplexVertex = SimplexVertex(x.copyOf(), fx)
}

/**
 * Minimize a function using the Nelder-Mead algorithm.
 */
fun nelderMead(
    f: (DoubleArray) -> Double,
    x0: DoubleArray,
    options: OptimizeOptions = defaultOptions()
): OptimizeResult {
    val n = x0.size
    var functionCalls = 0

    // Create initial simplex
    val simplex = mutableListOf<SimplexVertex>()
    simplex.add(SimplexVertex(x0.copyOf(), f(x0)))
    functionCalls++

    for (i in 0 until n) {
        val h = INITIAL_SIMPLEX_SCALE * max(abs(x0[i]), 1.0)
        val vertex = x0.copyOf()
        vertex[i] += h
        simplex.add(SimplexVertex(vertex, f(vertex)))
        functionCalls++
    }

    var iteration = 0

    while (iteration < options.maxIterations) {
        // Sort vertices by function value (ascending)
        simplex.sortBy { it.fx }

        val fBest = simplex[0].fx
        val fWorst = simplex[n].fx
        val fSecondWorst = simplex[n - 1].fx

        // Check convergence: function value spread
        val fValues = simplex.map { it.fx }.toDoubleArray()
        val fStd = standardDeviation(fValues)

        if (fStd < options.funcTol) {
            return OptimizeResult(
                x = simplex[0].x.copyOf(),
                `fun` = fBest,
                gradient = null,
                iterations = iteration,
                functionCalls = functionCalls,
                gradientCalls = 0,
                converged = true,
                message = "Converged: simplex function spread %.2e below tolerance".format(fStd)
            )
        }

        // Check convergence: simplex diameter (max distance from best vertex)
        var diameter = 0.0
        for (i in 1..n) {
            val d = VecOps.normInf(VecOps.sub(simplex[i].x, simplex[0].x))
            if (d > diameter) diameter = d
        }

        if (diameter < options.stepTol) {
            return OptimizeResult(
                x = simplex[0].x.copyOf(),
                `fun` = fBest,
                gradient = null,
                iterations = iteration,
                functionCalls = functionCalls,
                gradientCalls = 0,
                converged = true,
                message = "Converged: simplex diameter %.2e below tolerance".format(diameter)
            )
        }

        iteration++

        val best = simplex[0]
        val worst = simplex[n]

        // Compute centroid of all vertices except worst
        val centroid = simplex[0].x.copyOf()
        for (i in 1 until n) {
            for (j in 0 until n) {
                centroid[j] += simplex[i].x[j]
            }
        }
        for (j in 0 until n) {
            centroid[j] /= n.toDouble()
        }

        // Reflection: x_r = centroid + alpha * (centroid - worst)
        val reflected = VecOps.addScaled(centroid, VecOps.sub(centroid, worst.x), ALPHA)
        val fReflected = f(reflected)
        functionCalls++

        if (fReflected < fSecondWorst && fReflected >= fBest) {
            // Accept reflection
            simplex[n] = SimplexVertex(reflected, fReflected)
            continue
        }

        if (fReflected < fBest) {
            // Try expansion: x_e = centroid + gamma * (reflected - centroid)
            val expanded = VecOps.addScaled(centroid, VecOps.sub(reflected, centroid), GAMMA)
            val fExpanded = f(expanded)
            functionCalls++

            if (fExpanded < fReflected) {
                simplex[n] = SimplexVertex(expanded, fExpanded)
            } else {
                simplex[n] = SimplexVertex(reflected, fReflected)
            }
            continue
        }

        // Contraction
        if (fReflected < fWorst) {
            // Outside contraction: x_c = centroid + rho * (reflected - centroid)
            val contracted = VecOps.addScaled(centroid, VecOps.sub(reflected, centroid), RHO)
            val fContracted = f(contracted)
            functionCalls++

            if (fContracted <= fReflected) {
                simplex[n] = SimplexVertex(contracted, fContracted)
                continue
            }
        } else {
            // Inside contraction: x_c = centroid + rho * (worst - centroid)
            val contracted = VecOps.addScaled(centroid, VecOps.sub(worst.x, centroid), RHO)
            val fContracted = f(contracted)
            functionCalls++

            if (fContracted < fWorst) {
                simplex[n] = SimplexVertex(contracted, fContracted)
                continue
            }
        }

        // Shrink: move all vertices towards the best
        for (i in 1..n) {
            val newX = VecOps.add(best.x, VecOps.scale(VecOps.sub(simplex[i].x, best.x), SIGMA))
            simplex[i] = SimplexVertex(newX, f(newX))
            functionCalls++
        }
    }

    // Max iterations reached
    return OptimizeResult(
        x = simplex[0].x.copyOf(),
        `fun` = simplex[0].fx,
        gradient = null,
        iterations = iteration,
        functionCalls = functionCalls,
        gradientCalls = 0,
        converged = false,
        message = "Stopped: reached maximum iterations (${options.maxIterations})"
    )
}

private fun standardDeviation(values: DoubleArray): Double {
    if (values.isEmpty()) return 0.0
    val mean = values.average()
    val variance = values.map { (it - mean) * (it - mean) }.average()
    return sqrt(variance)
}
